src/06_ffi.md

```diff
@@ -228,17 +207,316 @@ library offers portable type aliases in `std:os::raw` (or `core::os::raw`):
+respect to the number of possible bit patterns of the same size. Mishandling an
+`enum` value provided by a foreign code may lead to type confusion and have
+severe consequences on software security. Unfortunately, checking an `enum`
+value at the FFI boundary is not simple on both sides.
+
+On the Rust side, it consists to actually use an integer type in the `extern`
+block declaration, a *robust* type, and then to perform a checked conversion
+to the enum type.
+
+On the foreign side, it is possible only if the other language allows for
+stricter checks than plain C. `enum class` in C++ are for instance allowable.
+Note however that as for reference the actual `extern "C"` ABI of
+`enum class` is implementation defined and should be verified for each
+environment.
+
+> ### Recommendation {{#check FFI-NOENUM | Do not use incoming Rust `enum` at FFI boundary}}
+>
+> In a secure Rust development, when interfacing with a foreign language,
+> the Rust code should not accept incoming values from Rust `enum` types.
+>
+> Exceptions include Rust `enum` types that are:
+>
+> - opaque in the foreign language and only manipulated from the
+>   Rust side,
+> - bound to safe enums in the foreign language, e.g. `enum class` types in C++.
+
+Concerning fieldless enums, crates like [`num_derive`] or [`num_enum`] allows
+developer to easily provide safe conversion from integer to enumeration and may
+be use to safely convert a integer (provided from a C `enum`) into a Rust enum.
+
+[num_derive]: https://crates.io/crates/num_derive
+[num_enum]: https://crates.io/crates/num_enum
+
+### Opaque types
+
+Opacifying types is a good way to increase modularity in software development.
+When doing multilingual development, it is something very common.
+
+> ### Recommendation {{#check FFI-R-OPAQUE | Use dedicated Rust types for foreign opaque types}}
+>
+> In a secure Rust development, when binding foreign opaque types, one should
+> use pointers to dedicated opaque types rather than `c_void` pointers.
+
+Currently the recommended way to make a foreign opaque type is like so:
+
+```rust ignore
+#[repr(C)]
+pub struct Foo {_private: [u8; 0]}
+extern {
+    pub fn foo(arg: *mut Foo);
+}
+```
+
+The not yet implemented [RFC 1861] proposes to facilitate the coding by allowing
+to declare opaque types in `extern` blocks.
+
+[RFC 1861]: https://rust-lang.github.io/rfcs/1861-extern-types.html
+
+> ### Recommendation {{#check FFI-C-OPAQUE | Use incomplete C/C++ `struct` pointers to make type opaque}}
+>
+> In a secure Rust development, when interfacing with C or C++, Rust types that
+> are to be considered opaque in C/C++ should be translated as incomplete
+> `struct` type (i,e., declared without definition) and be provided with
+> a dedicated constructor and destructor.
+
+Example of opaque Rust type:
+
+```rust
+struct Opaque {
+    // (...) details to be hidden
+}
+
+#[no_mangle]
+unsafe extern fn new_opaque() -> *mut Opaque {
+    Box::into_raw(Box::new(Opaque {
```

Comment:
catch unwind and return `ptr::null_mut()` ?

Reply:
Yes!